src/core/Akka.Streams/Dsl/RestartFlow.cs

```diff
@@ -301,25 +301,35 @@ protected RestartWithBackoffLogic(
         protected SubSinkInlet<TOut> CreateSubInlet(Outlet<TOut> outlet)
         {
             var sinkIn = new SubSinkInlet<TOut>(this, $"RestartWithBackoff{_name}.subIn");
-
             sinkIn.SetHandler(new LambdaInHandler(
-                onPush: () => Push(Out, sinkIn.Grab()),
+                onPush: () =>
+                {
+                    Push(Out, sinkIn.Grab());
+                },
                 onUpstreamFinish: () =>
                 {
-                    if (_finishing || MaxRestartsReached() || _onlyOnFailures)
+                    if (_finishing || (_onlyOnFailures && !MaxRestartsReached()))
```

Comment:
`A_restart_with_backoff_flow_should_not_restart_on_completion_when_maxRestarts_is_reached` has been racy because we never checked to see if `_finishing` was true during a normal completion event - we'd wait for another restart attempt instead.